File: sdks/go/intractsdk/sdk.go
package intractsdk

import (
	"fmt"
	"strings"
)

type Contract struct {
	Scope    string
	Intent   string
	Priority int
	Domain   string
	Input    []string
	Output   []string
	Effect   []string
	Forbid   []string
	Require  []string
	Validate []string
	Meaning  string
}

func csv(values []string) string {
	return strings.Join(values, ",")
}

func Inline(c Contract, prefix string) string {
	if prefix == "" {
		prefix = "//"
	}
	priority := c.Priority
	if priority == 0 {
		priority = 3
	}

	parts := []string{
		"@intract.v1",
		"scope:" + c.Scope,
		"intent:" + c.Intent,
		fmt.Sprintf("priority:%d", priority),
	}

	if c.Domain != "" {
		parts = append(parts, "domain:"+c.Domain)
	}
	if len(c.Input) > 0 {
		parts = append(parts, "input:"+csv(c.Input))
	}
	if len(c.Output) > 0 {
		parts = append(parts, "output:"+csv(c.Output))
	}
	if len(c.Effect) > 0 {
		parts = append(parts, "effect:"+csv(c.Effect))
	}
	if len(c.Forbid) > 0 {
		parts = append(parts, "forbid:"+csv(c.Forbid))
	}
	if len(c.Require) > 0 {
		parts = append(parts, "require:"+csv(c.Require))
	}
	if len(c.Validate) > 0 {
		parts = append(parts, "validate:"+csv(c.Validate))
	}
	if c.Meaning != "" {
		parts = append(parts, fmt.Sprintf("meaning:%q", c.Meaning))
	}

	return prefix + " " + strings.Join(parts, " ")
}

File: src/intract/sdk.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ContractBuilder:
    scope: str
    intent: str
    priority: int = 3
    domain: str = ""
    input: tuple[str, ...] = ()
    output: tuple[str, ...] = ()
    effect: tuple[str, ...] = ()
    forbid: tuple[str, ...] = ()
    require: tuple[str, ...] = ()
    validate: tuple[str, ...] = ()
    meaning: str = ""

    def to_inline(self, prefix: str = "#") -> str:
        parts = [
            "@intract.v1",
            f"scope:{self.scope}",
            f"intent:{self.intent}",
            f"priority:{self.priority}",
        ]
        if self.domain:
            parts.append(f"domain:{self.domain}")
        if self.input:
            parts.append("input:" + ",".join(self.input))
        if self.output:
            parts.append("output:" + ",".join(self.output))
        if self.effect:
            parts.append("effect:" + ",".join(self.effect))
        if self.forbid:
            parts.append("forbid:" + ",".join(self.forbid))
        if self.require:
            parts.append("require:" + ",".join(self.require))
        if self.validate:
            parts.append("validate:" + ",".join(self.validate))
        if self.meaning:
            parts.append(f'meaning:"{self.meaning}"')
        return f"{prefix} " + " ".join(parts)


def contract(**kwargs) -> ContractBuilder:
    return ContractBuilder(**kwargs)
